base/experiments/ad9910_ram.py

```diff
@@ -0,0 +1,331 @@
+            amplitude.
+        ramp_interval: float, the time interval between each step of the RAM
+            mode playback. Keep the interval at a multiple of 4*T_sysclk
+            (4*1 ns).
+        ram_type: RAMType, see the RAMType enum.
+        ram_mode: int, the playback mode of the RAM. AD9910 allows the
+            following RAM playback modes.
+
+            - RAM_MODE_DIRECTSWITCH:
+                Only the first data in the RAM profile is fed to the DDS
+
+            The other modes can fetch new RAM data to the DDS. The main
+            difference is the behavior once the last data is fetched.
+
+            - RAM_MODE_RAMPUP:
+                DDS will not fetch anything after getting the last data.
+            - RAM_MODE_BIDIR_RAMP:
+                Same as RAM_MODE_RAMPUP. However, an additional ramp down mode
+                is supported by setting profile=1.
+            - RAM_MODE_CONT_BIDIR_RAMP:
+                The RAM profile will ramp down after getting the last data.
+                Then ramp up after getting the first data, and so on.
+            - RAM_MODE_CONT_RAMPUP:
+                The RAM profile will repeat itself after getting the last data
+                from the RAM.
+
+            See the following sections on the datasheet regarding the exact
+            behavior of these 5 modes.
+            - RAM Direct Switch Mode (p.33)
+            - RAM Ramp-Up Mode (p.34, Figure 43)
+            - RAM Bidirectional Ramp Mode (p.38, Figure 46)
+            - RAM Continuous Bidirectional Ramp Mode (p.39, Figure 47)
+            - RAM Continuous Recirculate Mode (p.40, Figure 48)
+        base_frequency: float, the unmodulated DDS frequency.
+            0.0 Hz by default. This argument is optional.
+        base_phase: float, the unmodulated DDS phase.
+            0.0 turns by default. This argument is optional.
+        base_amplitude: float, the unmodulated DDS amplitude.
+            1.0 by default. This argument is optional.
+
+    Raises:
+        NotImplementedError: Unsupported RAM types found.
+    """
+    RAM_SIZE = 1024
+
+    def __init__(self, dds, data, ramp_interval, ram_type, ram_mode,
+                 base_frequency=0, base_phase=0, base_amplitude=1.0):
+        # Make sure the RAM can hold the entire sequence
+        assert (len(data) <= RAMProfile.RAM_SIZE)
+
+        ram = np.zeros((len(data),), dtype=np.int32)
+        # Avoid contaminating the passed in data list
+        data = data.copy()
+        # SPI transaction of AD9910 mandates the transmission of higher
+        # order words before lower order words
+        # The data is reversed such that the first word shows up first
+        data.reverse()
+
+        # Initialize default values for the FTW, POW, and ASF registers
+        # This is with accordance to set_mu() in AD9910
+        self.ftw, self.pow, self.asf = 0, 0, 0x3fff
+
+        # Encode the FTW, POW, ASF, and raw data en masse
+        match ram_type:
```

Comment:
Is `match` only in python 3.10? If so, we may still want to use `if` here to keep compatibility with some of the control computers that are on 3.8/3.9.